ratelimit/http_test.go

```diff
@@ -0,0 +1,407 @@
+	received int64
+}
+
+func (m *mockSharedLimiter) GetRateLimits(ctx context.Context, req *request) (remaining, resetTime int64, err error) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	if req.limit > m.received {
+		m.received++
+		return m.received, mockResetTime, nil
+	}
+
+	return m.received, mockResetTime, errOverLimit
+}
+
+func (m *mockSharedLimiter) reset() {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	m.received = 0
+}
+
+// nolint:dupl,funlen,scopelint
+func TestWithSharedRateLimiter(t *testing.T) {
+	matcherOne, matcherTwo := compileMatchers(t)
+	cases := []struct {
+		name      string
+		reqNum    int
+		configs   []Config
+		pathTests []pathTestParams
+	}{
+		{
+			name:   "no config",
+			reqNum: 5,
+			pathTests: []pathTestParams{
+				{
+					path:                    testPathOne,
+					waitBetween:             1 * time.Millisecond,
+					expectedOK:              5,
+					expectedTooManyRequests: 0,
+				},
+			},
+		},
+		{
+			name: "one rate limiter",
+			configs: []Config{
+				{
+					Tenant:  testTenant,
+					Matcher: matcherOne,
+					Limit:   1,
+					Window:  10 * time.Second,
+				},
+			},
+			reqNum: 5,
+			pathTests: []pathTestParams{
+				{
+					path:                    testPathOne,
+					waitBetween:             1 * time.Millisecond,
+					expectedOK:              1,
+					expectedTooManyRequests: 4,
+				},
+			},
+		},
+		{
+			name: "two rate limiters",
+			configs: []Config{
+				{
+					Tenant:  testTenant,
+					Matcher: matcherOne,
+					Limit:   1,
+					Window:  10 * time.Second,
+				},
+				{
+					Tenant:  testTenant,
+					Matcher: matcherTwo,
+					Limit:   3,
+					Window:  10 * time.Second,
+				},
+			},
+			reqNum: 5,
+			pathTests: []pathTestParams{
+				{
+					path:                    testPathOne,
+					waitBetween:             1 * time.Millisecond,
+					expectedOK:              1,
+					expectedTooManyRequests: 4,
+				},
+				{
+					path:                    testPathTwo,
+					waitBetween:             1 * time.Millisecond,
+					expectedOK:              3,
+					expectedTooManyRequests: 2,
+				},
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			limiter := &mockSharedLimiter{}
+			logger := logger.NewLogger("debug", "logfmt", "observatorium")
+			rlmw := WithSharedRateLimiter(logger, limiter, c.configs...)
+
+			r := chi.NewMux()
+
+			r.Group(func(r chi.Router) {
+				r.Use(authentication.WithTenant)
+				r.Use(rlmw)
+
+				r.HandleFunc(testPathOne+"/{tenant}", func(res http.ResponseWriter, req *http.Request) {
+					res.WriteHeader(http.StatusOK)
+				})
+				r.HandleFunc(testPathTwo+"/{tenant}", func(res http.ResponseWriter, req *http.Request) {
+					res.WriteHeader(http.StatusOK)
+				})
+			})
+
+			ts := httptest.NewServer(r)
+
+			for _, pathTest := range c.pathTests {
+				limiter.reset()
+				gotOKs, gotTooManyRequests, gotHeaders := launchTestRequests(t, ts.URL, pathTest, c.reqNum)
+
+				if pathTest.expectedOK != gotOKs {
+					t.Fatalf(
```

Comment:
It might be cleaner if we use https://pkg.go.dev/github.com/efficientgo/tools/core@v0.0.0-20210731122119-5d4a0645ce9a/pkg/testutil  - not a blocker, let's try  from next one (: